README.md
<div align="center">
<h1>dartsult</h1>
</div>
<div align="center">

`dartsult` is a library for developers who want to handle error in Rust style.

English | [简体中文](README-zh_CN.md)

[<img alt="github" src="https://img.shields.io/badge/GITHUB-dartsult-8da0cb?style=for-the-badge&logo=Github" height="22">][Github-url]
[<img alt="pub package" src="https://img.shields.io/pub/v/dartsult.svg?style=for-the-badge&logo=dart&color=5ab5f0" height="22">][package-url]

[<img alt="Build" src="https://img.shields.io/badge/Build-passing-brightgreen?style=for-the-badge&logo=Github-Actions" height="22">][CI-url]
[<img alt="codecov" src="https://img.shields.io/codecov/c/gh/al8n/dartsult?style=for-the-badge&logo=codecov" height="22">][codecov-url]

[<img alt="lisence" src="https://img.shields.io/badge/License-Apache%202.0-blue.svg?style=for-the-badge&logo=Apache" height="22">][license-url]



</div>

## Usage

A simple usage example:

```dart
import 'package:dartsult/dartsult.dart';

class MockException implements Exception {
  final String msg;

  MockException(this.msg);

  @override
  String toString() {
    return 'MockException($msg)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is MockException && other.msg == msg;
  }

  @override
  // TODO: implement hashCode
  int get hashCode => msg.hashCode;
}

Future<Result<Void, MockException>> voidResult(bool shouldPanic) async {
  return await mockLogic(null, 'cannot get void value',
      shouldPanic: shouldPanic)
      .then(
        (value) => Result.ok<Void, MockException>(Void()),
    onError: (error, stackTrace) =>
        Result.error<Void, MockException>(
          error,
          stackTrace: stackTrace,
        ),
  );
}

Future<Result<int, MockException>> intResult(bool shouldPanic) async {
  return await mockLogic<int>(5, 'cannot get int value',
      shouldPanic: shouldPanic)
      .then(
        (value) => Result.ok<int, MockException>(value),
    onError: (error, stackTrace) => Result.error<int, MockException>(
      error,
      stackTrace: stackTrace,
    ),
  );
}

/// mockLogic is used to mock some network I/O, file I/O or other asynchronous
/// logic
Future<T> mockLogic<T>(T val, String msg, {bool shouldPanic = false}) {
  return Future.delayed(Duration(milliseconds: 100), () {
    if (shouldPanic) {
      throw MockException(msg);
    }
    return val;
  });
}

void main() async {

  MockException mock = MockException('cannot get int value');
  MockException mock1 = MockException('cannot get void value');
  MockException mock2 = MockException('from another error');

  Result intRst = await intResult(false);
  assert(intRst.isOk());
  assert(intRst.unwrap() == 5);
  assert(intRst.contains(5));
  assert(intRst.contains(6) == false);

  Result<int, MockException> intErrRst = await intResult(true);
  assert(intErrRst.isError(), true);
  assert(intErrRst.unwrapError().toString() ==
      'MockException(cannot get int value)');
  assert(intErrRst.containsError(mock));
  assert(intErrRst.unwrapOr(6) == 6);
  assert(intErrRst.unwrapOrElse(() => 7) == 7);

  Result voidRst = await voidResult(false);
  assert(voidRst.isOk());
  assert(voidRst.unwrap() == Void());
  assert(voidRst.contains(Void()));

  Result voidErrRst = await voidResult(true);
  assert(voidErrRst.isError(), true);
  assert(voidErrRst.unwrapError().toString() ==
      'MockException(cannot get void value)');
  assert(voidErrRst.containsError(mock1));
  assert(voidErrRst.unwrapOr(Void()) == Void());
  assert(voidErrRst.unwrapOrElse(() => Void()) == Void());
}
```

## TODO
1. Implements `Option` class

2. Implements below methods for `Result`
    - [ ] `Option<T> ok()` 
    - [ ] `Option<E> err()`
    - [ ] `iter`
   
3. Implements `Either` class

#### License

<sup>
Licensed under <a href="LICENSE">Apache License, Version
2.0</a>.
</sup>
<br>
<sub>
Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in this crate by you, as defined in the Apache-2.0 license.
</sub>

[Github-url]: https://github.com/al8n/dartsult/
[CI-url]: https://github.com/al8n/dartsult
[codecov-url]: https://app.codecov.io/gh/al8n/dartsult/
[license-url]: LICENSE
[package-url]: https://pub.dartlang.org/packages/dartsult 

CHANGELOG.md
## 0.1.3
- fix `meta` dependency conflict with `flutter_test`
## 0.1.2
- add detailed comments for methods in `Result`
- add more test cases
- rename `core.dart` to `result.dart`
## 0.1.1
- remove some documents
- update README
## 0.1.0
- add `map`,`mapError`, `mapOr` and `mapOrElse` methods
- reconstruct `Result` to `Result<T, E>` 
- private `Ok<T>` and `Error<T>`

## 0.0.1

- Initial version.


test/result_test.dart
import 'package:dartsult/src/result.dart';
import 'package:test/test.dart';

class MockException implements Exception {
  final String msg;

  MockException(this.msg);

  @override
  String toString() {
    return 'MockException($msg)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is MockException && other.msg == msg;
  }

  @override
  int get hashCode => msg.hashCode;
}

Future<Result<Void, MockException>> voidResult(bool shouldPanic) async {
  return await mockLogic(null, 'cannot get void value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<Void, MockException>(Void()),
    onError: (error, stackTrace) => Result.error<Void, MockException>(
      error,
    ),
  );
}

Future<Result<int, MockException>> intResult(bool shouldPanic) async {
  return await mockLogic<int>(5, 'cannot get int value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<int, MockException>(value),
    onError: (error, stackTrace) => Result.error<int, MockException>(
      error,
    ),
  );
}

/// mockLogic is used to mock some network I/O, file I/O or other asynchronous
/// logic
Future<T> mockLogic<T>(T val, String msg, {bool shouldPanic = false}) {
  return Future.delayed(Duration(milliseconds: 100), () {
    if (shouldPanic) {
      throw MockException(msg);
    }
    return val;
  });
}

void main() {
  group('Result tests', () {
    MockException mock = MockException('cannot get int value');

    MockException mock1 = MockException('cannot get void value');

    MockException mock2 = MockException('from another error');

    test('Test int result with ok', () async {
      Result intRst = await intResult(false);
      expect(intRst.isOk(), true);
      expect(intRst.unwrap(), 5);
      expect(intRst.contains(5), true);
      expect(intRst.contains(6), false);
    });

    test('Test int result with error', () async {
      Result<int, MockException> intErrRst = await intResult(true);
      expect(intErrRst.isError(), true);
      expect(intErrRst.unwrapError().toString(),
          'MockException(cannot get int value)');
      expect(intErrRst.containsError(mock), true);
      expect(intErrRst.unwrapOr(6), 6);
      expect(intErrRst.unwrapOrElse((_) => 7), 7);
    });

    test('Test void result with ok', () async {
      Result voidRst = await voidResult(false);
      expect(voidRst.isOk(), true);
      expect(voidRst.unwrap(), Void());
      expect(voidRst.contains(Void()), true);
    });

    test('Test void result with error', () async {
      Result voidErrRst = await voidResult(true);
      expect(voidErrRst.isError(), true);
      expect(voidErrRst.unwrapError().toString(),
          'MockException(cannot get void value)');
      expect(voidErrRst.containsError(mock1), true);
      expect(voidErrRst.unwrapOr(Void()), Void());
      expect(voidErrRst.unwrapOrElse((_) => Void()), Void());
    });

    test('Test unwrap', () {
      Result<int, String> x = Result.ok(2);
      expect(x.unwrap(), 2);

      x = Result.error("emergency failure");
      try {
        x.unwrap();
      } catch (e) {
        expect(e.toString(), "Null check operator used on a null value");
      }
    });

    test('Test unwrapError', () {
      Result<int, String> x = Result.error("emergency failure");
      expect(x.unwrapError(), "emergency failure");

      x = Result.ok(2);
      try {
        x.unwrapError();
      } catch (e) {
        expect(e.toString(), "Null check operator used on a null value");
      }
    });

    test('Test unwrapOr', () {
      int defaultt = 2;
      Result<int, String> x = Result.ok(9);
      expect(x.unwrapOr(defaultt), 9);

      x = Result.error("error");
      expect(x.unwrapOr(defaultt), defaultt);
    });

    test('Test unwrapOrElse', () {
      int count(String x) {
        return x.length;
      }

      expect(Result.ok<int, String>(2).unwrapOrElse(count), 2);
      expect(Result.error<int, String>("foo").unwrapOrElse(count), 3);
    });

    test('Test map', () {
      String stringify(int x) {
        return 'code: $x';
      }

      Result<int, String> x = Result.ok<int, String>(13);
      expect(x.map(stringify), Result.ok<String, String>('code: 13'));
    });

    test('Test mapError', () async {
      String stringify(int x) {
        return 'error code: $x';
      }

      Result<int, int> x = Result.ok(2);
      expect(x.mapError(stringify), Result.ok<int, String>(2));

      x = Result.error(13);
      expect(
          x.mapError(stringify), Result.error<int, String>("error code: 13"));
    });

    test('Test mapOr', () {
      Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
      int val = x.mapOr<int>(42, (v) => v.length);
      expect(val, 3);

      Result<String, dynamic> y = Result.error<String, dynamic>("bar");
      int val1 = y.mapOr<int>(42, (v) => v.length);
      expect(val1, 42);
    });

    test('Test mapOrElse', () {
      int k = 21;

      Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
      int val = x.mapOrElse<int>((v) => k * 2, (v) => v.length);
      expect(val, 3);

      Result<String, dynamic> y = Result.error<String, dynamic>("bar");
      int val1 = y.mapOrElse<int>((v) => k * 2, (v) => v.length);
      expect(val1, 42);
    });

    test('Test or', () {
      Result<int, String> x = Result.ok(2);
      Result<int, String> y = Result.error("late error");
      expect(x.or(y), Result.ok<int, String>(2));

      x = Result.error("early error");
      y = Result.ok(2);
      expect(x.or(y), Result.ok<int, String>(2));

      x = Result.ok(2);
      y = Result.ok(100);
      expect(x.or(y), Result.ok<int, String>(2));
    });

    test('Test orElse', () {
      Result<int, int> sq(int x) {
        return Result.ok(x * x);
      }

      Result<int, int> err(int x) {
        return Result.error(x);
      }

      expect(
          Result.ok<int, int>(2).orElse(sq).orElse(sq), Result.ok<int, int>(2));
      expect(Result.ok<int, int>(2).orElse(err).orElse(sq),
          Result.ok<int, int>(2));
      expect(Result.error<int, int>(3).orElse(sq).orElse(err),
          Result.ok<int, int>(9));
      expect(Result.error<int, int>(3).orElse(err).orElse(err),
          Result.error<int, int>(3));
    });

    test('Test and', () {
      Result<int, String> x = Result.ok(2);
      Result<String, String> y = Result.error("late error");

      expect(x.and(y), Result.error<String, String>("late error"));

      x = Result.error("early error");
      y = Result.ok("foo");
      expect(x.and(y), Result.error<String, String>("early error"));

      x = Result.error("not a 2");
      y = Result.error("late error");
      expect(x.and(y), Result.error<String, String>("not a 2"));

      x = Result.ok(2);
      y = Result.ok("different result type");
      expect(x.and(y), Result.ok<String, String>("different result type"));
    });

    test('Test andThen', () {
      Result<int, int> sq(int x) {
        return Result.ok(x * x);
      }

      Result<int, int> err(int x) {
        return Result.error(x);
      }

      expect(Result.ok<int, int>(2).andThen(sq).andThen(sq),
          Result.ok<int, int>(16));
      expect(Result.ok<int, int>(2).andThen(sq).andThen(err),
          Result.error<int, int>(4));
      expect(Result.ok<int, int>(2).andThen(err).andThen(sq),
          Result.error<int, int>(2));
      expect(Result.error<int, int>(3).andThen(sq).andThen(sq),
          Result.error<int, int>(3));
    });

    test('Test Ok toString, hashCode', () async {
      Result<int, String> okint = Result.ok(5);
      Result<int, String> okint1 = Result.ok(5);
      expect(okint.toString(), 'Ok(5)');
      expect(okint, okint1);
      expect(okint.hashCode, "Ok".hashCode + 5.hashCode);

      Result<int, String> errint = Result.error("asd");
      expect(errint.hashCode, 'Error'.hashCode + 'asd'.hashCode);
    });

    test('Test Void to hashcode', () async {
      expect(Void().hashCode, 0);
    });
  });
}


test/dartsult_test.dart
import 'package:dartsult/dartsult.dart';
import 'package:dartsult/src/core.dart';
import 'package:test/test.dart';

class MockException implements Exception {
  final String msg;

  MockException(this.msg);

  @override
  String toString() {
    return 'MockException($msg)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is MockException && other.msg == msg;
  }

  @override
  int get hashCode => msg.hashCode;
}

Future<Result<Void, MockException>> voidResult(bool shouldPanic) async {
  return await mockLogic(null, 'cannot get void value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<Void, MockException>(Void()),
    onError: (error, stackTrace) =>
        Result.error<Void, MockException>(
          error,
        ),
  );
}

Future<Result<int, MockException>> intResult(bool shouldPanic) async {
  return await mockLogic<int>(5, 'cannot get int value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<int, MockException>(value),
    onError: (error, stackTrace) => Result.error<int, MockException>(
      error,
    ),
  );
}

/// mockLogic is used to mock some network I/O, file I/O or other asynchronous
/// logic
Future<T> mockLogic<T>(T val, String msg, {bool shouldPanic = false}) {
  return Future.delayed(Duration(milliseconds: 100), () {
    if (shouldPanic) {
      throw MockException(msg);
    }
    return val;
  });
}

void main() {
  group('dartsult tests', () {
    MockException mock = MockException('cannot get int value');

    MockException mock1 = MockException('cannot get void value');

    MockException mock2 = MockException('from another error');

    test('Test int result with ok', () async {
      Result intRst = await intResult(false);
      expect(intRst.isOk(), true);
      expect(intRst.unwrap(), 5);
      expect(intRst.contains(5), true);
      expect(intRst.contains(6), false);
    });

    test('Test int result with error', () async {
      Result<int, MockException> intErrRst = await intResult(true);
      expect(intErrRst.isError(), true);
      expect(intErrRst.unwrapError().toString(),
          'MockException(cannot get int value)');
      expect(intErrRst.containsError(mock), true);
      expect(intErrRst.unwrapOr(6), 6);
      expect(intErrRst.unwrapOrElse(() => 7), 7);
    });

    test('Test void result with ok', () async {
      Result voidRst = await voidResult(false);
      expect(voidRst.isOk(), true);
      expect(voidRst.unwrap(), Void());
      expect(voidRst.contains(Void()), true);
    });

    test('Test void result with error', () async {
      Result voidErrRst = await voidResult(true);
      expect(voidErrRst.isError(), true);
      expect(voidErrRst.unwrapError().toString(),
          'MockException(cannot get void value)');
      expect(voidErrRst.containsError(mock1), true);
      expect(voidErrRst.unwrapOr(Void()), Void());
      expect(voidErrRst.unwrapOrElse(() => Void()), Void());
    });

    test('Test map', () {
      Result<int, String> x = Result.ok<int, String>(1);
      var v = x.map<String>((p0) => Result.ok('foo'));
      expect(v.unwrap(), 'foo');
    });

    test('test mapError', () async {
      var x = await intResult(true);
      expect(x.mapError((err) => err.msg).toString(), 'Error(${mock.msg})');
    });

    test('Test mapOr', () {
      Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
      int val = x.mapOr<int>(42, (v) => v.length);
      expect(val, 3);

      Result<String, dynamic> y = Result.error<String, dynamic>("bar");
      int val1 = y.mapOr<int>(42, (v) => v.length);
      expect(val1, 42);
    });

    test('Test mapOrElse', () {
      int k = 21;

      Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
      int val = x.mapOrElse<int>((v) => k * 2, (v) => v.length);
      expect(val, 3);

      Result<String, dynamic> y = Result.error<String, dynamic>("bar");
      int val1 = y.mapOrElse<int>((v) => k * 2, (v) => v.length);
      expect(val1, 42);

    });

    test('Test Ok toString, hashCode', () async {
      Result<int, String> okint = Result.ok(5);
      Result<int, String> okint1 = Result.ok(5);
      expect(okint.toString(), 'Ok(5)');
      expect(okint, okint1);
      expect(okint.hashCode, "Ok".hashCode + 5.hashCode);

      Result<int, String> errint = Result.error("asd");
      expect(errint.hashCode, 'Error'.hashCode + 'asd'.hashCode);
    });

    test('Test Void to hashcode', () async {
      expect(Void().hashCode, 0);
    });
  });
}


lib/src/core.dart
import 'package:meta/meta.dart';

enum ResultType {
  ok,
  error,
}

/// [Ok(T)], representing success and containing a value.
class _Ok<T> {
  final T? val;

  _Ok({this.val});

  @override
  String toString() {
    return 'Ok(${val?.toString()})';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }

    return other is _Ok<T> && other.val == val;
  }

  @override
  int get hashCode => "Ok".hashCode + val.hashCode;
}

/// [Error(E)], representing error and containing an error value.
class _Error<E> {
  final E? val;

  /// Default constructor
  const _Error({this.val});

  @override
  String toString() {
    return 'Error(${val?.toString()})';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }

    return other is _Error<E> && other.val == val;
  }

  @override
  int get hashCode => "Error".hashCode + val.hashCode;
}

/// Void is a class for function returns a void.
@immutable
class Void {
  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    return other is Void;
  }

  /// All [Void] instance should equal.
  @override
  int get hashCode => 0;
}

/// `Result` is a type that represents either success ([`ok`]) or failure ([`error`]).
@immutable
class Result<T, E> {
  final ResultType type;
  
  final dynamic rawValue;

  /// Contains the success value
  final _Ok<T> _ok;

  /// Contains the failure value
  final _Error<E> _error;

  const Result._({
    required _Ok<T> ok,
    required _Error<E> error,
    required this.type,
    required this.rawValue,
  })  : _ok = ok,
        _error = error;

  static Result<T, E> ok<T, E>(T val) {
    return Result._(ok: _Ok<T>(val: val), error: _Error<E>(), type: ResultType.ok, rawValue: val);
  }

  static Result<T, E> error<T, E>(E err) {
    return Result._(ok: _Ok<T>(), error: _Error<E>(val: err), type: ResultType.error, rawValue: err);
  }

  /// Returns `true` if the result is [ok].
  bool isOk() => type == ResultType.ok;

  /// Returns `true` if the result is [error].
  bool isError() => type == ResultType.error;

  /// Returns `true` if the result is an [`ok`] value containing the given value.
  bool contains(T val) {
    if (_ok.val != null) {
      return _Ok(val: val) == _ok;
    }

    return false;
  }

  /// Returns `true` if the result is an [`error`] value containing the given value.
  bool containsError(E err) {
    if (_error.val != null) {
      return _Error(val: err) == _error;
    }

    return false;
  }

  /// Returns the contained [ok] value.
  ///
  /// Because this function may panic, its use is generally discouraged.
  /// Instead, prefer to use pattern matching and handle the [error]
  /// case explicitly, or call [unwrapOr] or [unwrapOrElse].
  T unwrap() => _ok.val!;

  /// Returns the contained [error] value.
  E unwrapError() => _error.val!;

  /// Returns the contained [`ok`] value or a provided default.
  ///
  /// Arguments passed to `unwrapOr` are eagerly evaluated; if you are passing
  /// the result of a function call, it is recommended to use [unwrapOrElse],
  /// which is lazily evaluated.
  T unwrapOr(T instance) => _ok.val != null ? _ok.val! : instance;

  /// Returns the contained [ok] value or computes it from a closure.
  T unwrapOrElse(T Function() builder) => _ok.val != null ? _ok.val! : builder();

  /// Maps a `Result<T, E>` to `Result<U, E>` by applying a function to a
  /// contained [ok] value, leaving an [error] value untouched.
  ///
  /// This function can be used to compose the results of two functions.
  Result<U, E> map<U>(Result<U, E> Function(T) op) {
    if (_ok.val != null) {
      return op(_ok.val!);
    }
    return Result.error(_error.val!);
  }

  /// Maps a `Result<T, E>` to `Result<T, F>` by applying a function to a
  /// contained [error] value, leaving an [ok] value untouched.
  ///
  /// This function can be used to pass through a successful result while handling
  /// an error.
  Result<T, F> mapError<F>(F Function(E err) op) {
    if (_ok.val != null) {
      return Result.ok(_ok.val!);
    }
    return Result.error(op(_error.val!));
  }

  /// Returns the provided default (if [error]), or
  /// applies a function to the contained value (if [ok]),
  ///
  /// Arguments passed to `map_or` are eagerly evaluated; if you are passing
  /// the result of a function call, it is recommended to use [mapOrElse],
  /// which is lazily evaluated.
  U mapOr<U>(U defaultt, U Function(T) op) {
    if (_ok.val != null) {
      return op(_ok.val!);
    }
    return defaultt;
  }

  /// Maps a `Result<T, E>` to `U` by applying a fallback function to a
  /// contained [error] value, or a default function to a
  /// contained [ok] value.
  ///
  /// This function can be used to unpack a successful result
  /// while handling an error.
  ///
  U mapOrElse<U>(U Function(E err) defaultt, U Function(T val) op) {
    if (_ok.val != null) {
      return op(_ok.val!);
    }
    return defaultt(_error.val!);
  }

  @override
  int get hashCode {
    switch(type) {
      case ResultType.ok:
        return _ok.hashCode;
      case ResultType.error:
        return _error.hashCode;
    }
  }

  @override
  String toString() {
    switch (type){
      case ResultType.ok:
        return _ok.toString();
      case ResultType.error:
        return _error.toString();
    }
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }

    if (other.runtimeType != runtimeType) {
      return false;
    }

    if (other is Result<T, E>) {
      if (type != other.type) {
        return false;
      }
      return rawValue == other.rawValue;
    }

    return false;
  }
}


example/dartsult_example.dart
import 'package:dartsult/dartsult.dart';

class MockException implements Exception {
  final String msg;

  MockException(this.msg);

  @override
  String toString() {
    return 'MockException($msg)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is MockException && other.msg == msg;
  }

  @override
  int get hashCode => msg.hashCode;
}

Future<Result<Void, MockException>> voidResult(bool shouldPanic) async {
  return await mockLogic(null, 'cannot get void value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<Void, MockException>(Void()),
    onError: (error, stackTrace) => Result.error<Void, MockException>(
      error,
    ),
  );
}

Future<Result<int, MockException>> intResult(bool shouldPanic) async {
  return await mockLogic<int>(5, 'cannot get int value',
          shouldPanic: shouldPanic)
      .then(
    (value) => Result.ok<int, MockException>(value),
    onError: (error, stackTrace) => Result.error<int, MockException>(
      error,
    ),
  );
}

/// mockLogic is used to mock some network I/O, file I/O or other asynchronous
/// logic
Future<T> mockLogic<T>(T val, String msg, {bool shouldPanic = false}) {
  return Future.delayed(Duration(milliseconds: 100), () {
    if (shouldPanic) {
      throw MockException(msg);
    }
    return val;
  });
}

Result<int, int> sq(int x) {
  return Result.ok(x * x);
}

Result<int, int> err(int x) {
  return Result.error(x);
}

void unwrap() {
  Result<int, String> x = Result.ok(2);
  assert(x.unwrap() == 2);

  x = Result.error("emergency failure");
  try {
    x.unwrap();
  } catch (e) {
    assert(e.toString() == "Null check operator used on a null value");
  }
}

void unwrapError() {
  Result<int, String> x = Result.error("emergency failure");
  assert(x.unwrapError() == "emergency failure");

  x = Result.ok(2);
  try {
    x.unwrapError();
  } catch (e) {
    assert(e.toString() == "Null check operator used on a null value");
  }
}

void unwrapOr() {
  int defaultt = 2;
  Result<int, String> x = Result.ok(9);
  assert(x.unwrapOr(defaultt) == 9);

  x = Result.error("error");
  assert(x.unwrapOr(defaultt) == defaultt);
}

void unwrapOrElse() {
  int count(String x) {
    return x.length;
  }

  assert(Result.ok<int, String>(2).unwrapOrElse(count) == 2);
  assert(Result.error<int, String>("foo").unwrapOrElse(count) == 3);
}

void or() {
  Result<int, String> x = Result.ok(2);
  Result<int, String> y = Result.error("late error");
  assert(x.or(y) == Result.ok<int, String>(2));

  x = Result.error("early error");
  y = Result.ok(2);
  assert(x.or(y) == Result.ok<int, String>(2));

  x = Result.ok(2);
  y = Result.ok(100);
  assert(x.or(y) == Result.ok<int, String>(2));
}

void orElse() {
  assert(
      Result.ok<int, int>(2).orElse(sq).orElse(sq) == Result.ok<int, int>(2));
  assert(
      Result.ok<int, int>(2).orElse(err).orElse(sq) == Result.ok<int, int>(2));
  assert(Result.error<int, int>(3).orElse(sq).orElse(err) ==
      Result.ok<int, int>(9));
  assert(Result.error<int, int>(3).orElse(err).orElse(err) ==
      Result.error<int, int>(3));
}

void and() {
  Result<int, String> x = Result.ok(2);
  Result<String, String> y = Result.error("late error");

  assert(x.and(y) == Result.error<String, String>("late error"));

  x = Result.error("early error");
  y = Result.ok("foo");
  assert(x.and(y) == Result.error<String, String>("early error"));

  x = Result.error("not a 2");
  y = Result.error("late error");
  assert(x.and(y) == Result.error<String, String>("not a 2"));

  x = Result.ok(2);
  y = Result.ok("different result type");
  assert(x.and(y) == Result.ok<String, String>("different result type"));
}

void andThen() {
  assert(Result.ok<int, int>(2).andThen(sq).andThen(sq) ==
      Result.ok<int, int>(16));
  assert(Result.ok<int, int>(2).andThen(sq).andThen(err) ==
      Result.error<int, int>(4));
  assert(Result.ok<int, int>(2).andThen(err).andThen(sq) ==
      Result.error<int, int>(2));
  assert(Result.error<int, int>(3).andThen(sq).andThen(sq) ==
      Result.error<int, int>(3));
}

void map() {
  String stringify(int x) {
    return 'code: $x';
  }

  Result<int, String> x = Result.ok<int, String>(13);
  assert(x.map(stringify) == Result.ok<String, String>('code: 13'));
}

void mapError() {
  String stringify(int x) {
    return 'error code: $x';
  }

  Result<int, int> x = Result.ok(2);
  assert(x.mapError(stringify) == Result.ok<int, String>(2));

  x = Result.error(13);
  assert(x.mapError(stringify) == Result.error<int, String>("error code: 13"));
}

void mapOr() {
  Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
  int val = x.mapOr<int>(42, (v) => v.length);
  assert(val == 3);

  Result<String, dynamic> y = Result.error<String, dynamic>("bar");
  int val1 = y.mapOr<int>(42, (v) => v.length);
  assert(val1 == 42);
}

void mapOrElse() {
  int k = 21;

  Result<dynamic, String> x = Result.ok<dynamic, String>("foo");
  int val = x.mapOrElse<int>((v) => k * 2, (v) => v.length);
  assert(val == 3);

  Result<String, dynamic> y = Result.error<String, dynamic>("bar");
  int val1 = y.mapOrElse<int>((v) => k * 2, (v) => v.length);
  assert(val1 == 42);
}

main() async {
  MockException mock = MockException('cannot get int value');
  MockException mock1 = MockException('cannot get void value');
  MockException mock2 = MockException('from another error');

  Result intRst = await intResult(false);
  assert(intRst.isOk());
  assert(intRst.unwrap() == 5);
  assert(intRst.contains(5));
  assert(intRst.contains(6) == false);

  Result<int, MockException> intErrRst = await intResult(true);
  assert(intErrRst.isError(), true);
  assert(intErrRst.unwrapError().toString() ==
      'MockException(cannot get int value)');
  assert(intErrRst.containsError(mock));
  assert(intErrRst.unwrapOr(6) == 6);
  assert(intErrRst.unwrapOrElse((_) => 7) == 7);

  Result voidRst = await voidResult(false);
  assert(voidRst.isOk());
  assert(voidRst.unwrap() == Void());
  assert(voidRst.contains(Void()));

  Result voidErrRst = await voidResult(true);
  assert(voidErrRst.isError(), true);
  assert(voidErrRst.unwrapError().toString() ==
      'MockException(cannot get void value)');
  assert(voidErrRst.containsError(mock1));
  assert(voidErrRst.unwrapOr(Void()) == Void());
  assert(voidErrRst.unwrapOrElse((_) => Void()) == Void());

  // unwrap
  unwrap();

  // unwrapError
  unwrapError();

  // unwrapOr
  unwrapOr();

  // unwrapOrElse
  unwrapOrElse();

  // or
  or();

  // orElse example
  orElse();

  // and
  and();

  // andThen
  andThen();

  // map
  map();

  // mapError
  mapError();

  // mapOr
  mapOr();

  // mapOrElse
  mapOrElse();
}
